Add integer overflow checks for large compressed tensor dimensions and nnz

pearu opened the pull request:
With the previous PR allowing large compressed tensors (dimensions larger than `2 ** 31 - 1`), sparse compressed tensor invariants checks may give false-positive results:
```python
>>> nnz=2**31
>>> torch.sparse.check_sparse_tensor_invariants.enable()
>>> torch.sparse_csr_tensor(torch.arange(nnz+1, dtype=torch.int32), torch.zeros(nnz, dtype=torch.int32), torch.ones(nnz), (nnz, 1))
tensor(crow_indices=tensor([          0,           1,           2,  ...,
                             2147483646,  2147483647, -2147483648]),
       col_indices=tensor([0, 0, 0,  ..., 0, 0, 0]),
       values=tensor([1., 1., 1.,  ..., 1., 1., 1.]), size=(2147483648, 1),
       nnz=2147483648, layout=torch.sparse_csr)
```
(notice that the last entry in `crow_indices` is invalid) or raise a bogus exception as in
```python
>>> torch.sparse_csr_tensor(torch.arange(nnz+1, dtype=torch.int32), torch.arange(nnz, dtype=torch.int32), torch.ones(nnz), (nnz, 1))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: `0 <= col_indices < ncols` is not satisfied.
```
(notice that `col_indices` is actually valid).

This PR fixes the above-reported bugs by introducing integer overflow checks for sparse compressed tensors dimensions as well as nnz.


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102613
* __->__ #102530
* #102523

pearu committed:
Add integer overflow checks for large compressed tensor dimensions and nnz

[ghstack-poisoned]

nikitaved commented on aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:
```diff
@@ -240,6 +240,29 @@ void _validate_compressed_sparse_indices_kernel(
   // For TensorIterator's output: no void lambdas.
   const auto dummy = at::empty({1}, cidx.options());
 
+  // Catch integer overflow from large dimensions. Otherwise, the
+  // invariant checks may fail with bogus exceptions or succeed with
+  // false-positive results when int64_t typed dimensions are cast to
+  // index dtype that corresponds to smaller interger type such as
+  // int32_t.
+  {
+    AT_DISPATCH_INDEX_TYPES(idx.scalar_type(), NAME, [cdim, dim, nnz]() {
+      if (cdim_name == CDimName::CRow) {
```
nit: probably would make sense to put that at the very top?

pearu replied:
yes, we can do that, although, it does not matter much because the tests above use int64 comparisons and there exist no integer overflow issues.

nikitaved reviewed (approved):
LGTM! Thank you, @pearu !

pearu committed:
Update on "Add integer overflow checks for large compressed tensor dimensions and nnz"

With the previous PR allowing large compressed tensors (dimensions larger than `2 ** 31 - 1`), sparse compressed tensor invariants checks may give false-positive results:
```python
>>> nnz=2**31
>>> torch.sparse.check_sparse_tensor_invariants.enable()
>>> torch.sparse_csr_tensor(torch.arange(nnz+1, dtype=torch.int32), torch.zeros(nnz, dtype=torch.int32), torch.ones(nnz), (nnz, 1))
tensor(crow_indices=tensor([          0,           1,           2,  ...,
                             2147483646,  2147483647, -2147483648]),
       col_indices=tensor([0, 0, 0,  ..., 0, 0, 0]),
       values=tensor([1., 1., 1.,  ..., 1., 1., 1.]), size=(2147483648, 1),
       nnz=2147483648, layout=torch.sparse_csr)
```
(notice that the last entry in `crow_indices` is invalid) or raise a bogus exception as in
```python
>>> torch.sparse_csr_tensor(torch.arange(nnz+1, dtype=torch.int32), torch.arange(nnz, dtype=torch.int32), torch.ones(nnz), (nnz, 1))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: `0 <= col_indices < ncols` is not satisfied.
```
(notice that `col_indices` is actually valid).

This PR fixes the above-reported bugs by introducing integer overflow checks for sparse compressed tensors dimensions as well as nnz.





[ghstack-poisoned]
